### DISCUSSION TITLE
Formal json specification

### DISCUSSION THREAD

**Discussion by flxo** [arendst/Tasmota#17827] [Q&A] [answered]
Hello,

I'd like to ask if there's a (complete?) json spec of the tasmota json model?  There are some [examples](https://tasmota.github.io/docs/JSON-Status-Responses/#basic-response) but I didn't find anything else.

Thanks,

@flxo

**Comment by barbudor** [comment, marked as answer]
Unfortunately not
Each driver may add it's piece and we are not using a framework to manage JSON which could automatically such neat documentation.

**Reply by flxo** [reply]
Thanks.

Hm. I implemented some parsing routines and can just test what my sensors emit. Helpful would be traces of large installations with lots of entities. Not sure if that exists publicly.

**Reply by barbudor** [reply]
If you can make a bot that crawls github issues and discussions.. 😂

**Reply by sfromis** [reply]
There is simply no "JSON model" at all, the payloads are intended to be reasonably easy to understand when you see them, within the general structure of most JSON having a number of objects, each having a number of key/value pairs. Sometimes, new values are added if a driver gains more function, and new sensors happens from time to time. Other places, new values are also occasionally added, with no perceived "need" for some type of formal specification of such straightforward data.

When it is sensors, parsing is really easy to do, just normal JSON within the same general structure, and the expectation is that subscribers to the data can ignore values they are not prepared to handle, focusing on what they need to use instead.

**Reply by flxo** [reply]
@sfromis Understood. I know that maintaining such a model is hard and no fun.

May I ask another question: Is there a definition of timestamp format? I found three different ones:
 
* `2023-01-30T14:21:58`
* `2018.02.01 21:29:40`
* `Thu Feb 15 00:00:50 2018`

**Reply by sfromis** [reply]
The first is within ISO 8601, where I normally set options to include milliseconds and time zone:
`2023-01-30T15:33:49.959+01:00`
Without time zone, it is local time, including user-selected adjustments from UTC.

The two others depends on where you saw them, I'm not expecting those in JSON payloads from Tasmota.

**Reply by barbudor** [reply]
The [Time](https://tasmota.github.io/docs/Commands/#time) command could impact the exact format of the date
I'm also using the format 1
![image](https://user-images.githubusercontent.com/2996042/215510292-0e709763-f050-4357-93f1-812048227fbf.png)

I don't know in which dev environment you are working but in Pythonn dateutil.parser.parse can handle quite a large number of variant format

**Reply by sfromis** [reply]
I'm using `Time 4` and `SetOption52 1` for the full format I posted. Parsing is easier when I do not have to worry about what "local time" means, when time zone may not be the same always.

**Reply by flxo** [reply]
> The two others depends on where you saw them, I'm not expecting those in JSON payloads from Tasmota.

`2018.02.01 21:29:40` and `Thu Feb 15 00:00:50 2018` are from your [project documentation](https://tasmota.github.io/docs/JSON-Status-Responses/#am2301). Probably just outdated.

**Reply by sfromis** [reply]
Badly outdated, I'd consider those very old examples useless for getting details of what can be generated, at *most* for getting an idea of the general structure, but even a quick glance shows that those timestamp formats are not the only thing outside of what you can expect. I'm even wondering if the file has been damaged while being transferred from one docs format to another. I'm close to saying unusable.

**Reply by flxo** [reply]
haha. I derived some Rust structs from them and barely nothing worked.

Thanks for your support guys! I will just have a look what my setup publishes or dig in the codebase.

**Reply by sfromis** [reply]
Here's a current example of sensor payloads, with some details depending on my settings.
` xxxxxxxx/tele/SENSOR = {"Time":"2023-01-31T13:24:03.728+01:00","Switch1":"ON","BME280":{"Temperature":21.35,"Humidity":37.75,"DewPoint":6.33,"Pressure":1007.11,"SeaPressure":1008.72},"BH1750":{"Illuminance":32},"APDS9960":{"Red":31,"Green":33,"Blue":30,"Illuminance":22,"CCT":6897,"Proximity":53},"VEML7700":{"Illuminance":31,"WhiteContent":91},"MCP9808":{"Temperature":21.37},"PressureUnit":"hPa","TempUnit":"C"}`

**Comment by flxo** [comment]
Just out of curiosity: Is there a json lib that would fit into tasmota? I used to do `printf json` myself and know how painful that can be.

**Reply by arendst** [reply]
The JSON libs have been abandoned or minimized to keep code as small as. Possible simply because there is no room for features that can be handled by well programming too.

Sorry.

**Reply by sfromis** [reply]
With ESP32 and Berry, valid JSON can be generated from a map structure with key/value pairs (including maps within maps).
